tests/unit/services/userService.test.ts

```diff
@@ -1,119 +1,171 @@
-		await expect(userService.verifyUserEmail(confirmation.id.substring(0, confirmation.id.length - 1))).rejects.toThrow();
-		await expect(userService.verifyUserEmail(`${confirmation.id} `)).rejects.toThrow();
-
-		// Verify their email
-		const oldUser = { ...confirmation.user };
-		const newExpected = {
-			...oldUser,
-			accountStatus: AccountStatus.Verified
-		};
-		let user = await userService.verifyUserEmail(confirmation.id);
-		expect(user).toMatchObject(newExpected);
-		expect(user.profile).toBeNull();
-
-		// A second validation attempt should fail
-		await expect(userService.verifyUserEmail(confirmation.id)).rejects.toThrow();
-
-		// Attempt 2 invalid authenticates
-		await expect(userService.authenticate(fixture.email, 'incorrect')).rejects.toThrow();
-		await expect(userService.authenticate(fixture.email, `${fixture.password} `)).rejects.toThrow();
-
-		// Attempt a valid authenticate
-		await expect(userService.authenticate(fixture.email, fixture.password)).resolves.toMatchObject(newExpected);
-
-		// Attempt 2 invalid profile puts
-		await expect(userService.putUserProfile(user.id, {} as any)).rejects.toThrow();
-		await expect(userService.putUserProfile(user.id, { yearOfStudy: 1.5, course: 'Computer Science', profilePicture: 'asdf' })).rejects.toThrow();
-
-		// Set valid profile
-		const profileFixture = {
-			yearOfStudy: 1,
-			course: 'Computer Science',
-			facebook: 'student324'
-		};
-		user = await userService.putUserProfile(user.id, profileFixture);
-		expect(user.profile).toBeTruthy();
-		expect(user.profile).toMatchObject(profileFixture);
-
-		// Update profile again
-		const profileFixture2 = {
-			...profileFixture,
-			twitter: 'testacct',
-			profilePicture: '5327d0cc39d3047b1d3079fbb02bf11c'
-		};
-		user = await userService.putUserProfile(user.id, profileFixture2);
-		expect(user.profile).toBeTruthy();
-		expect(user.profile).toMatchObject({ ...profileFixture2, id: user.profile!.id });
-
-		// Update profile again, but this time try to change the id
-		// The service should ignore the id field
-		const oldId = user.profile!.id;
-		const profileFixture3 = {
-			...profileFixture2,
-			id: '0a0841a4-6f74-4e3b-85ac-1207727be375'
-		};
-		user = await userService.putUserProfile(user.id, profileFixture3);
-		expect(user.profile).toBeTruthy();
-		expect(user.profile).toMatchObject({ ...profileFixture2, id: oldId });
-	});
+	describe('registerUser', () => {
+		test('Registers a user with valid details', async () => {
+			const { email, forename, surname } = users[0];
+			const details = { email, forename, surname, password: 'thunderbolt' };
+			const confirmation = await userService.registerUser(details);
+			expect(confirmation.id).toBeTruthy();
+			expect(confirmation.user).toMatchObject({
+				email,
+				forename,
+				surname,
+				accountType: AccountType.User,
+				accountStatus: AccountStatus.Unverified
+			});
+			expect(await verifyPassword('thunderbolt', confirmation.user.password)).toStrictEqual(true);
```

Comment:
I would say if you just want to test `registerUser`, then try to avoid using `verifyPassword`. Instead you could hardcode the hashed+salted version of the password `thunderbolt` in order to compare it.

Since you are using `randomBytes` when hashing the password, I would say it would be best to mock out this randomness (by mocking `hashPassword`) so you control the return value.

The idea being that you could instead check the validity of the `hashPassword`/`verifyPassword` functions in `test/unit/util/password.test.ts`.

Reply:
Done!